src/CodeGen_ARM.cpp

```diff
@@ -149,133 +110,365 @@ CodeGen_ARM::CodeGen_ARM(Target target)
+
+    // arm64 expects a 32-bit constant.
+    {nullptr, "sqshrn.v8i8", Int(8, 8), "saturating_shift_right_narrow", {Int(16, 8), Int(32)}},
+    {nullptr, "uqshrn.v8i8", UInt(8, 8), "saturating_shift_right_narrow", {UInt(16, 8), Int(32)}},
+    {nullptr, "sqshrn.v4i16", Int(16, 4), "saturating_shift_right_narrow", {Int(32, 4), Int(32)}},
+    {nullptr, "uqshrn.v4i16", UInt(16, 4), "saturating_shift_right_narrow", {UInt(32, 4), Int(32)}},
+    {nullptr, "sqshrn.v2i32", Int(32, 2), "saturating_shift_right_narrow", {Int(64, 2), Int(32)}},
+    {nullptr, "uqshrn.v2i32", UInt(32, 2), "saturating_shift_right_narrow", {UInt(64, 2), Int(32)}},
+    {nullptr, "sqshrun.v8i8", UInt(8, 8), "saturating_shift_right_narrow", {Int(16, 8), Int(32)}},
+    {nullptr, "sqshrun.v4i16", UInt(16, 4), "saturating_shift_right_narrow", {Int(32, 4), Int(32)}},
+    {nullptr, "sqshrun.v2i32", UInt(32, 2), "saturating_shift_right_narrow", {Int(64, 2), Int(32)}},
+
+    // Saturating doubling multiply keep high half.
+    {"vqdmulh.v4i16", "sqdmulh.v4i16", Int(16, 4), "qdmulh", {Int(16, 4), Int(16, 4)}},
+    {"vqdmulh.v2i32", "sqdmulh.v2i32", Int(32, 2), "qdmulh", {Int(32, 2), Int(32, 2)}},
+
+    {"vqdmulh.v8i16", "sqdmulh.v8i16", Int(16, 8), "qdmulh", {Int(16, 8), Int(16, 8)}},
+    {"vqdmulh.v4i32", "sqdmulh.v4i32", Int(32, 4), "qdmulh", {Int(32, 4), Int(32, 4)}},
+
+    // Saturating doubling multiply keep high half with rounding.
+    {"vqrdmulh.v4i16", "sqrdmulh.v4i16", Int(16, 4), "qrdmulh", {Int(16, 4), Int(16, 4)}},
+    {"vqrdmulh.v2i32", "sqrdmulh.v2i32", Int(32, 2), "qrdmulh", {Int(32, 2), Int(32, 2)}},
+
+    {"vqrdmulh.v8i16", "sqrdmulh.v8i16", Int(16, 8), "qrdmulh", {Int(16, 8), Int(16, 8)}},
+    {"vqrdmulh.v4i32", "sqrdmulh.v4i32", Int(32, 4), "qrdmulh", {Int(32, 4), Int(32, 4)}},
+};
+// clang-format on
+
+}  // namespace
+
+void CodeGen_ARM::init_module() {
+    CodeGen_Posix::init_module();
+
+    if (neon_intrinsics_disabled()) {
+        return;
     }
-}
 
-Value *CodeGen_ARM::call_pattern(const Pattern &p, llvm::Type *t, const vector<llvm::Value *> &args) {
-    if (target.bits == 32) {
-        return call_intrin(t, p.intrin_lanes, p.intrin32, args);
-    } else {
-        return call_intrin(t, p.intrin_lanes, p.intrin64, args);
+    std::string prefix = target.bits == 32 ? "llvm.arm.neon." : "llvm.aarch64.neon.";
+    for (const ArmIntrinsic &i : intrinsic_defs) {
+        const char *intrin_name = nullptr;
+        if (target.bits == 32) {
+            intrin_name = i.arm32;
+        } else {
+            intrin_name = i.arm64;
+        }
+        if (!intrin_name) {
+            continue;
+        }
+        std::string full_name = intrin_name;
+        if (!starts_with(full_name, "llvm.")) {
+            full_name = prefix + full_name;
+        }
+
+        Type ret_type = i.ret_type;
+        std::vector<Type> arg_types;
+        arg_types.reserve(4);
```

Comment:
I'd be more comfortable with: 4 -> sizeof(i.arg_types) / sizeof(i.arg_types[0])

Or some other way to synchronize this 4 with the 4 in the struct definition.